x/mm2/keeper/keeper.go

```diff
@@ -0,0 +1,158 @@
+package keeper
+
+import (
+	"fmt"
+
+	"cosmossdk.io/collections"
+	"cosmossdk.io/core/store"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/skip-mev/slinky/x/mm2/types"
+)
+
+// Keeper is the module's keeper implementation.
+type Keeper struct {
+	cdc codec.BinaryCodec
+
+	// module authority
+	authority sdk.AccAddress
+
+	// markets is keyed by CurrencyPair string (BASE/QUOTE) and contains
+	// the list of all Markets.
+	markets collections.Map[types.TickerString, types.Market]
+
+	// lastUpdated is the last block height the marketmap was updated.
+	lastUpdated collections.Item[uint64]
+
+	// params is the module's parameters.
+	params collections.Item[types.Params]
+}
+
+// NewKeeper initializes the keeper and its backing stores.
+func NewKeeper(ss store.KVStoreService, cdc codec.BinaryCodec, authority sdk.AccAddress) *Keeper {
+	sb := collections.NewSchemaBuilder(ss)
+
+	// Create the collections item that will track the module parameters.
+	params := collections.NewItem[types.Params](
+		sb,
+		types.ParamsPrefix,
+		"params",
+		codec.CollValue[types.Params](cdc),
+	)
+
+	return &Keeper{
+		cdc:         cdc,
+		authority:   authority,
+		markets:     collections.NewMap(sb, types.MarketsPrefix, "markets", types.TickersCodec, codec.CollValue[types.Market](cdc)),
+		lastUpdated: collections.NewItem[uint64](sb, types.LastUpdatedPrefix, "last_updated", types.LastUpdatedCodec),
+		params:      params,
+	}
+}
+
+// SetLastUpdated sets the lastUpdated field to the current block height.
+func (k *Keeper) SetLastUpdated(ctx sdk.Context, height uint64) error {
+	return k.lastUpdated.Set(ctx, height)
+}
+
+// GetLastUpdated gets the last block-height the market map was updated.
+func (k *Keeper) GetLastUpdated(ctx sdk.Context) (uint64, error) {
+	return k.lastUpdated.Get(ctx)
+}
+
+// GetMarket returns a market from the store by its currency pair string ID.
+func (k *Keeper) GetMarket(ctx sdk.Context, tickerStr string) (types.Market, error) {
+	return k.markets.Get(ctx, types.TickerString(tickerStr))
+}
+
+// GetAllMarkets returns the set of Market objects currently stored in state
+// as a map[TickerString] -> Markets.
+func (k *Keeper) GetAllMarkets(ctx sdk.Context) (map[string]types.Market, error) {
```

Comment:
how does this return type differ from `collections.Map[types.TickerString, types.Market]`

Reply:
So, a `collections.Map` stores data in the KVstore in a map-like fashion. The access pattern is a `Get` which returns a single item, or to iterate through the collection which would let you return a list, or map.  Here, we return a map.